Update Custom Voice API to a new version (2024-02-01-preview).

ForrestGumb opened the pull request:
# Data Plane API - Pull Request
Update Custom Voice API to a new version (2024-02-01-preview).
Compare with previous version (2023-12-01-preview), only one change. Added failureReason property in Consent.

I missed failure reason in initial version.
Create consent is a long-running-operation. When it fails, customer need a way to know the failure reason.
It should return failure reason in the same way as Model and PersonalVoice.
So, I added failureReason in Consent.properties follow the same patter as Model and PersonalVoice.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

Commit:
Update Custom Voice API to a new version (2024-02-01-preview).
Compare to previous version (2023-12-01-preview), only one change. Added failureReason property in Consent.

mikekistler commented:
Approving breaking changes because [the breaking change check failed with a runtime exception](https://github.com/Azure/azure-sdk-tools/issues/7278). Manual inspection of the diff shows no breaking changes.

weidongxu-microsoft commented:
/azp run

ForrestGumb commented on specification/cognitiveservices/data-plane/Speech/TextToSpeech/preview/2024-02-01-preview/texttospeech.json:
```diff
@@ -0,0 +1,2630 @@
+        "projectId": {
+          "$ref": "#/definitions/ResourceId"
+        },
+        "properties": {
+          "$ref": "#/definitions/ConsentProperties"
+        },
```
No strong reason.
In Custom Voice schema definition, we put shared properties (like id, name, description, status, createdDateTime) in top level, and put type specific properties under ‘properties’. Properties under ‘properties’ are optional.
It’s too late to move them up. Other object type (Endpoint, Model, TrainingSet, PersonalVoice) already followed this design.

mikekistler replied:
Properties that are nested for no good reason are a poor user experience. I'll approve for this preview but please fix this for your next preview.

weidongxu-microsoft commented on specification/cognitiveservices/data-plane/Speech/TextToSpeech/preview/2024-02-01-preview/texttospeech.json:
```diff
@@ -0,0 +1,2630 @@
+          }
+        }
+      },
+      "delete": {
```
One lint error seems to be here.

It should contain a `produces` as well (the error response is still a json).

ForrestGumb replied:
Accepted. Added produces in delete method. Now LintDiff error is gone.

weidongxu-microsoft commented on specification/cognitiveservices/data-plane/Speech/TextToSpeech/preview/2024-02-01-preview/texttospeech.json:
```diff
@@ -0,0 +1,2630 @@
+    "ConsentFailureReason": {
+      "description": "Consent creation failure reason",
+      "enum": [
+        "None",
```
Is this `None` mean "consent does not fail" or "we don't know why it fails"?

ForrestGumb replied:
The 2nd one. consent failed, but we don't know why.

weidongxu-microsoft replied:
If service decide to keep the `None`, you may consider at least add `description` part in [x-ms-enum](https://azure.github.io/autorest/extensions/#x-ms-enum) to document it.

Commit:
Fix LintDiff error on CustomVoice API by adding produces on delete method.

weidongxu-microsoft commented:
There is some warning in LintDiff

1. we'd prefer service support [x-ms-error-code](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors) before GA
2. request with mutlipart/form-data should be used with caution

mikekistler reviewed (approved):
Looks good for preview. 👍

Commit:
Update Custom Voice API according to review feedback.
1. Remove 'None' from modelFailureReason and consentFailureReason.
2. Add x-ms-error-code in header

ForrestGumb committed:
Merge branch 'main' into junwg/CNV-API-7-3

ForrestGumb commented:
> There is some warning in LintDiff
> 
> 1. we'd prefer service support [x-ms-error-code](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors) before GA
> 2. request with multipart/form-data should be used with caution

1. Updated swagger to support x-ms-error-code.
2. I'm aware that multipart/form-data is not recommended. So, our API followed standard, use PUT to create with Azure blob URL as audio input. But for Consent and PersonalVoice, a common scenario is creation with small data. We removed dependency on Azure Blob API, supported additional POST create API with multipart/form-data input. This is reviewed with Mike Kistler and Jeffrey Richter.

weidongxu-microsoft commented:
````
Operation response '/responses/ErrorResponse' produces type couldn't be resolved
```
in LintDiff appears to be a false alert.

Commit:
Response shared "reponses" because LintDiff will report error on that.

Commit:
Fix markdown grammar in description.
